pkg/checker/functools.go

```diff
@@ -0,0 +1,113 @@
+	if err := validateFunction(f, args...); err != nil {
+		return nil, fmt.Errorf("invalid function: %v", err)
+	}
+	argValues := make([]reflect.Value, 0, len(args))
+	for _, arg := range args {
+		argValues = append(argValues, reflect.ValueOf(arg))
+	}
+	return func() error {
+		ret := reflect.ValueOf(f).Call(argValues)
+		// len(ret) is assumed to be 1 and of type `error`, because of type
+		// `CheckRunner`
+		if ret[0].IsNil() {
+			return nil
+		}
+		return ret[0].Interface().(error)
+	}, nil
+}
+
+func makeRemediationRunner(f interface{}, args ...interface{}) (RemediationRunner, error) {
+	if err := validateFunction(f, args...); err != nil {
+		return nil, fmt.Errorf("invalid function: %v", err)
+	}
+	argValues := make([]reflect.Value, 0, len(args))
+	for _, arg := range args {
+		argValues = append(argValues, reflect.ValueOf(arg))
+	}
+	return func() error {
+		ret := reflect.ValueOf(f).Call(argValues)
+		// len(ret) is assumed to be 1 and of type `error`, because of type
+		// `RemediationRunner`
+		if ret[0].IsNil() {
+			return nil
+		}
+		return ret[0].Interface().(error)
+	}, nil
+}
+
+// validateFunction checks whether `f` is a function, and whether
+// the passed parameters have the correct number and types.
+func validateFunction(f interface{}, args ...interface{}) error {
+	typ := reflect.TypeOf(f)
+	errorType := reflect.TypeOf((*error)(nil)).Elem()
+	if kind := typ.Kind(); kind != reflect.Func {
+		return fmt.Errorf("invalid check runner, want %v, got %v", reflect.Func, kind)
+	}
+	// check output parameter type(s). The function must return an error
+	// TODO: assess if we want to also allow (value, error) as acceptable return
+	//       types. This means also changing how values are returned in
+	//       `makeCheckRunner` and `makeRemediationRunner`.
+	switch {
+	case typ.NumOut() == 1:
+		// return type must be error
+		if typ.Out(0) != errorType {
+			return fmt.Errorf("invalid return type, want %v, got %v", errorType, typ.Out(0))
+		}
+	default:
+		return fmt.Errorf("invalid number of returned parameters, want 1, got %d", typ.NumOut())
+	}
+	// check input parameter types
+	if typ.IsVariadic() {
+		if len(args) < typ.NumIn()-1 {
+			return fmt.Errorf("wrong number of parameters for variadic function, want at least %d, got %d", typ.NumIn()-1, len(args))
+		}
+		// check non-variadic arguments
+		for idx := 0; idx < typ.NumIn()-1; idx++ {
+			want, got := typ.In(idx).Kind(), reflect.TypeOf(args[idx]).Kind()
+			if want != got {
+				return fmt.Errorf("wrong non-variadic parameter type at index %d: want %v, got %v", idx, want, got)
+			}
+		}
+		// check variadic arguments
+		vargs := typ.In(typ.NumIn() - 1)
+		if vargs.Kind() != reflect.Slice {
+			return fmt.Errorf("invalid variadic arguments: want %v, got %v", reflect.Slice, vargs.Kind())
+		}
+		for idx := typ.NumIn() - 1; idx < len(args); idx++ {
+			argType := reflect.TypeOf(args[idx])
+			if argType.Kind() == reflect.Interface {
+				if !argType.Implements(vargs.Elem()) {
```

Comment:
**error:**

If `vargs.Elem()` is not interface, it will panic.

Reply:
uh.. this is obvious (and I found that in the code) but not documented, good catch

Reply:
ah no I see. I thought that `Implements` would panic, not that `Elem()` would panic. And that's documented

Reply:
Sorry, I just explain myself very unclear. I meant `Implements` will panic when `vargs.Elem()` is not an interface.

Reply:
this should be covered by unit tests (still a couple of fixes and tests missing and coming soon)